src/internal/task/task_threads.go

```diff
@@ -0,0 +1,265 @@
+//go:build scheduler.threads
+
+package task
+
+import (
+	"sync/atomic"
+	"unsafe"
+)
+
+// If true, print verbose debug logs.
+const verbose = false
+
+// Scheduler-specific state.
+type state struct {
+	// Goroutine ID. The number here is not really significant and after a while
+	// it could wrap around. But it is useful for debugging.
+	id uintptr
+
+	// Thread ID, pthread_t or similar (typically implemented as a pointer).
+	thread threadID
+
+	// Highest address of the stack. It is stored when the goroutine starts, and
+	// is needed to be able to scan the stack.
+	stackTop uintptr
+
+	// Next task in the activeTasks queue.
+	QueueNext *Task
+
+	// Semaphore to pause/resume the thread atomically.
+	pauseSem Semaphore
```

Comment:
Shouldn't `Semaphore` be used with pointer receiver?

Reply:
No. The `state` struct (or actually the `Task` struct where it is embedded) will always be accessed via a pointer, like `*Task`. So essentially when calling a function like `pauseSem.Post()`, what happens is something like `SemaphorePost(&t.state.pauseSem)`.

Also, I think you mean to say that `pauseSem` should have the type `*Semaphore`? Since the methods on the `Semaphore` type do in fact have a pointer receiver:

```go
func (s *Semaphore) Post() { ... }
func (s *Semaphore) Wait() { ... }
```

Reply:
My point was about a pointer (pun intended :smile:) on `*Semaphore`. While `Semaphore` methods do have pointer receiver, the instances of `pauseSem` and `gcSem` members would get copied when `state` struct instance is ever passed around.

I've created playground snippet which uses stdlib `sync.Mutex` as an example to even trigger `go vet` warnings: https://go.dev/play/p/cV7h16_JS1R

Reply:
> While `Semaphore` methods do have pointer receiver, the instances of `pauseSem` and `gcSem` members would get copied when `state` struct instance is ever passed around.

That's why it should never be copied around!
Because it's an internal API, we can make sure this is the case. (And it is).